js/ui/codingManager.js

```diff
@@ -0,0 +1,638 @@
+
+const BUTTON_OFFSET_X = 50;
+const BUTTON_OFFSET_Y = 50;
+
+function animateBounce(actor) {
+    Tweener.removeTweens(actor);
+
+    Tweener.addTween(actor, {
+        scale_y: 1 + ICON_BOUNCE_MAX_SCALE,
+        scale_x: 1 + ICON_BOUNCE_MAX_SCALE,
+        translation_y: actor.height * ICON_BOUNCE_MAX_SCALE,
+        translation_x: actor.width * ICON_BOUNCE_MAX_SCALE / 2,
+        time: ICON_BOUNCE_ANIMATION_TIME * 0.25,
+        delay: 0.3,
+        transition: ICON_BOUNCE_ANIMATION_TYPE_1
+    });
+    Tweener.addTween(actor, {
+        scale_y: 1,
+        scale_x: 1,
+        translation_y: 0,
+        translation_x: 0,
+        time: ICON_BOUNCE_ANIMATION_TIME * 0.35,
+        transition: ICON_BOUNCE_ANIMATION_TYPE_2,
+        delay: ICON_BOUNCE_ANIMATION_TIME * 0.25 + 0.3,
+        onComplete: function() {animateBounce(actor);}
+    });
+}
+
+const CodingManager = new Lang.Class({
+    Name: 'CodingManager',
+
+    _init: function() {
+        this._sessions = [];
+        this._rotateInActors = [];
+        this._rotateOutActors = [];
+        this._firstFrameConnections = [];
+        this._watchdogId = 0;
+        this._codingApps = ['org.gnome.gedit', 'org.gnome.Weather'];
+    },
+
+    _isCodingApp: function(flatpakID) {
+        return this._codingApps.indexOf(flatpakID) != -1;
+    },
+
+    _isBuilder: function(flatpakID) {
+        return flatpakID === 'org.gnome.Builder';
+    },
+
+    addAppWindow: function(actor) {
+        let window = actor.meta_window;
+        if (!this._isCodingApp(window.get_flatpak_id()))
+            return;
+
+        this._addSwitcherToBuilder(actor);
+    },
+
+    addBuilderWindow: function(actor) {
+        let window = actor.meta_window;
+        if (!this._isBuilder(window.get_flatpak_id()))
+            return false;
+
+        this._cancelWatchdog();
+
+        let session = this._sessions[this._sessions.length-1];
+        if (session.actorBuilder)
+            return false;
+        let tracker = Shell.WindowTracker.get_default();
+        tracker.untrack_coding_app_window();
+
+        this._addSwitcherToApp(actor, session);
+        return true;
+    },
+
+    removeAppWindow: function(actor) {
+        let window = actor.meta_window;
+        if (!this._isCodingApp(window.get_flatpak_id()))
+            return;
+
+        this._removeSwitcherToBuilder(actor);
+    },
+
+    removeBuilderWindow: function(actor) {
+        let window = actor.meta_window;
+        if (!this._isBuilder(window.get_flatpak_id()))
+            return;
+
+        this._removeSwitcherToApp(actor);
+    },
+
+    _addSwitcherToBuilder: function(actorApp) {
+        let window = actorApp.meta_window;
+
+        let button = new St.Button({ style_class: 'view-source' });
+        let rect = window.get_frame_rect();
+        button.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        Main.layoutManager.addChrome(button);
```

Comment:
See my previous suggestion of factoring the code that creates a button into a separate function, and calling it from both here and `animateToBuilder`